miden-crypto/src/merkle/smt/large/mod.rs

```diff
@@ -426,6 +436,219 @@ impl<S: SmtStorage> LargeSmt<S> {
+    /// Returns the new root hash of the tree after applying all changes.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// - Any leaf would exceed [`MAX_LEAF_ENTRIES`](super::MAX_LEAF_ENTRIES) (1024 entries)
+    /// - Storage operations fail
+    ///
+    /// # Example
+    /// ```no_run
+    /// use miden_crypto::{
+    ///     EMPTY_WORD, Felt, Word,
+    ///     merkle::{LargeSmt, RocksDbConfig, RocksDbStorage},
+    /// };
+    ///
+    /// # fn main() -> Result<(), Box<dyn std::error::Error>> {
+    /// let storage = RocksDbStorage::open(RocksDbConfig::new("/path/to/db"))?;
+    /// let mut smt = LargeSmt::new(storage)?;
+    ///
+    /// let entries = vec![
+    ///     // Insert new entries
+    ///     (
+    ///         Word::new([Felt::new(1), Felt::new(0), Felt::new(0), Felt::new(0)]),
+    ///         Word::new([Felt::new(10), Felt::new(20), Felt::new(30), Felt::new(40)]),
+    ///     ),
+    ///     (
+    ///         Word::new([Felt::new(2), Felt::new(0), Felt::new(0), Felt::new(0)]),
+    ///         Word::new([Felt::new(11), Felt::new(22), Felt::new(33), Felt::new(44)]),
+    ///     ),
+    ///     // Delete an entry
+    ///     (Word::new([Felt::new(3), Felt::new(0), Felt::new(0), Felt::new(0)]), EMPTY_WORD),
+    /// ];
+    ///
+    /// let new_root = smt.insert_batch(entries)?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn insert_batch(
+        &mut self,
+        kv_pairs: impl IntoIterator<Item = (Word, Word)>,
+    ) -> Result<Word, LargeSmtError>
+    where
+        Self: Sized + Sync,
+    {
+        use rayon::prelude::*;
+
+        // Sort key-value pairs by leaf index
+        let mut sorted_kv_pairs: Vec<_> = kv_pairs.into_iter().collect();
+        sorted_kv_pairs.par_sort_by_key(|(key, _)| Self::key_to_leaf_index(key).value());
+
+        // Load leaves from storage
+        let (_leaf_indices, leaf_map) = self.load_leaves_for_pairs(&sorted_kv_pairs)?;
+
+        // Process leaves in parallel to get mutated leaves for tree building AND deltas
+        let (mut leaves, mutated_leaf_nodes, _new_pairs, leaf_count_delta, entry_count_delta) =
+            self.sorted_pairs_to_mutated_leaves_with_preloaded_leaves(sorted_kv_pairs, &leaf_map);
+
+        // Early return if no mutations
+        let old_root = self.root()?;
+        if leaves.is_empty() {
+            return Ok(old_root);
+        }
+
+        let mut loaded_subtrees: Map<NodeIndex, Option<Subtree>> = Map::new();
+
+        // Process each depth level in reverse, stepping by the subtree depth
+        for subtree_root_depth in
+            (0..=SMT_DEPTH - SUBTREE_DEPTH).step_by(SUBTREE_DEPTH as usize).rev()
+        {
+            // Build mutations and apply them to loaded subtrees
+            let (in_memory_mutations, mut subtree_roots, modified_subtrees) = leaves
+                .into_par_iter()
+                .map(|subtree_leaves| {
+                    let subtree_root_index = NodeIndex::new_unchecked(
```

Comment:
given the length of the function, this could be a function, iiuc it only takes the closure argument and `subtree_leaves` and `subtree_root_depth`